File: 卡码网/Js/test24.js
//24. 最长公共子序列
const rl=require("readline").createInterface({input:process.stdin});
const iter=rl[Symbol.asyncIterator]();
const readline=async ()=>(await iter.next()).value;

void async function(){
  let line;
  while((line=await readline())){
    const [s1,s2]=line.split(" ");
    let dp=new Array(s1.length+1);
    for(let i=0;i<s1.length+1;i++){
      dp[i]=new Array(s2.length+1).fill(0);
    }
    for(let i=1;i<s1.length+1;i++){
      for(let j=1;j<s2.length+1;j++){
        if(s1[i-1]==s2[j-1]) dp[i][j]=dp[i-1][j-1]+1;
        else dp[i][j]=Math.max(dp[i-1][j],dp[i][j-1]);
      }
    }
    console.log(dp[s1.length][s2.length]);
  }
}();

File: 卡码/KamaCoder/Python/test24.py
#24. 最长公共子序列
import sys

def main():
	for line in sys.stdin:
		[s1,s2]=line.split()
		dp=[[0 for x in range(len(s2)+1)] for x in range(len(s1)+1)]
		for i in range(1,len(s1)+1):
			for j in range(1,len(s2)+1):
				if s1[i-1]==s2[j-1]:
					dp[i][j]=dp[i-1][j-1]+1
				else:
					dp[i][j]=max(dp[i-1][j],dp[i][j-1])
		print(dp[len(s1)][len(s2)])

if __name__=="__main__":
	main()